File: evaluators/JSON-Interpreter.js

var fun_def_dicts = []
var functions = ['+', '-', '*', 'div', 'modulo', '=', '!=','<', '>', '<=', '>=']
var environment = []

function loadJSON(json_fun_defs) {
  for (var i = 0; i < json_fun_defs.length; i++) {
    fun_def_dicts[i] = JSON.parse(json_fun_defs[i]);
  }
  fs = fun_def_dicts.map(function (d) {
    return d.name;
  })
  functions = fs.concat(functions);
}

function main(forest, exp) {
  loadJSON(forest);
  return evaluate(exp);
}

function evaluate(d) {
  if('tag' in d) {
    var tag = d.tag;
    if (tag === 'number') {
      return Number(d.val);
    } else if (tag === 'identifier') {
      var name = d.name;
      if (name in environment[environment.length-1]) {
        return environment[environment.length-1][name];
      }
      else {
        console.log("Undefined Identifier" + name);
        return "Unidentified Identifier"
      }
    } else if (tag === 'case') {
      var cases = d.cases;
      for (var i = 0; i < cases.length; i++) {
        var case_exp = cases[i];
        var condition = evaluate(case_exp.condition);
        if (condition) return evaluate(case_exp.exp);
      }
      return evaluate(d.elseExp); //but what if the thing above returns?
    } else if (tag === 'call') {
      var fun = d.function;
      var argVals = d.argVals;
      var builtInFuns = ["+", '-', '*', 'div', 'modulo', '=', '!=','<', '>', '<=', '>='];
      if (builtInFuns.indexOf(fun.name) != -1) {
        var arg1 = evaluate(argVals[0]);
        var arg2 = evaluate(argVals[1]);
        var index = builtInFuns.indexOf(d.function.name)
        return builtIn(arg1, arg2, index);
      } else {
        //var args = [];
        /*for (var i = 0; i < argVals.length; i++) {
          arguments.push(evaluate(argVals[i]));
        }*/
        var args = argVals.map(function (arg) {return evaluate(arg);});
        if (functions.indexOf(fun.name) === -1) {
          return "ERROR: function is undefined" //how to do this nicer
        }
        var called_fun = fun_def_dicts[functions.indexOf(fun.name)];
        var called_fun_args = called_fun.args;
        var new_env = {};
      /*  for (var i = 0; i < called_fun_args.length; i++) {
          new_env[called_fun_args[i]] = arguments[i];
        }*/
        called_fun_args.forEach(function (cfa, i) {new_env[cfa] = args[i];});
        environment.push(new_env);
        var result = evaluate(called_fun.body);
        environment.pop();
        return result;
      }
    } else if (tag === 'list') {
      return d.list; //?
    } else if (tag === 'let') {
      var bindings = d.bindings;
      new_env = {};
      bindings.forEach(function (let_binding) {
        new_env[let_binding.name] = evaluate(let_binding.value);
      });
      environment.push(new_env);
      var result = evaluate(d.body);
      environment.pop();
      return result;
    }
  } else {
    return "ERROR: Unidentified object"
  }
}

function builtIn(a1, a2, i) {
  if (i === 0) return a1 + a2;
  else if (i === 1) return a1 - a2;
  else if (i === 2) return a1 * a2;
  else if (i === 3) return Math.floor(a1/a2);
  else if (i === 4) return a1 % a2;
  else if (i === 5) return a1 === a2;
  else if (i === 6) return a1 != a2;
  else if (i === 7) return a1 < a2;
  else if (i === 8) return a1 > a2;
  else if (i === 9) return a1 <= a2;
  else return a1 >= a2;
}

File: evaluators/JSON-Interpreter.py
import json

fun_def_dicts = []
functions = ['+', '-', '*', 'div', 'modulo', '=', '!=','<', '>', '<=', '>=']
environment = []


def loadJSON(json_fun_defs):
    global fun_def_dicts, functions
    fun_def_dicts = [json.loads(j) for j in json_fun_defs]
    fs = [d['name'] for d in fun_def_dicts]
    functions = fs + functions

def main(forest, exp):
    loadJSON(forest)
    if ('args' in exp): #if it's a function_def
        return evaluate(exp['body'])
    else:
        return evaluate(exp)


def evaluate(obj):
    global fun_def_dicts, functions, environment
    if ('tag' in obj):
        tag = obj['tag']
        if (tag == 'number'):
            return eval(obj['val'])
        elif (tag == 'identifier'):
            name = obj['name']
            if (name in environment[-1]):
                return environment[-1][name]
            else:
                return "Undefined identifier: %s" % name
        elif (tag == 'case'):
            cases = obj['cases']
            elseExp = obj['elseExp']
            for case_exp in cases:
                condition = evaluate(case_exp['condition'])
                if (condition):
                    return evaluate(case_exp['exp'])
            return evaluate(elseExp)
        elif (tag == 'call'):
            function = obj['function']
            argVals = obj['argVals']
            builtInFuns = ['+', '-', '*', 'div', 'modulo', '=', '!=','<', '>', '<=', '>=']
            if (function['name'] in builtInFuns):
                arg1 = evaluate(argVals[0])
                arg2 = evaluate(argVals[1])
                index = builtInFuns.index(obj['function']['name'])
                return builtIn(arg1, arg2, index)
            else:
                arguments = []
                for arg in argVals:
                    arguments.append(evaluate(arg))
                if (not (function['name'] in functions)):
                    return "ERROR: %s is undefined" % function['name']
                called_fun = fun_def_dicts[functions.index(function['name'])]
                called_fun_args = called_fun['args']
                new_env = {}
                for i in range(len(called_fun_args)):
                    new_env[called_fun_args[i]] = arguments[i]
                environment.append(new_env)
                result = evaluate(called_fun['body'])
                environment.pop()
                return result
        elif (tag == 'list'):
            #?
            #evaluate the inner items?
            return obj['list']
        elif (tag == 'let'):
            bindings = obj['bindings']
            new_env = {}
            for let_binding in bindings:
                new_env[let_binding['name']] = evaluate(let_binding['value'])
            environment.append(new_env)
            result = evaluate(obj['body'])
            environment.pop()
            return result
    else:
        return "Unidentified object"

def builtIn(a1, a2, i):
    if i == 0:
        return a1 + a2
    elif i == 1:
        return a1 - a2
    elif i == 2:
        return a1 * a2
    elif i == 3:
        return a1 // a2
    elif i == 4:
        return a1 % a2
    elif i == 5:
        return a1 == a2
    elif i == 6:
        return a1 != a2
    elif i == 7:
        return a1 < a2
    elif i == 8:
        return a1 > a2
    elif i == 9:
        return a1 <= a2
    else:
        return a1 >= a2
